Lib/fontTools/pens/ttGlyphPen.py

```diff
@@ -103,72 +215,73 @@ def closePath(self):
+class TTGlyphPointPen(TTGlyphBasePen, LogMixin, AbstractPointPen):
+    """
+    Pen used for drawing to a TrueType glyph.
 
-            component = GlyphComponent()
-            component.glyphName = glyphName
-            component.x, component.y = (otRound(v) for v in transformation[4:])
-            # quantize floats to F2Dot14 so we get same values as when decompiled
-            # from a binary glyf table
-            transformation = tuple(
-                floatToFixedToFloat(v, 14) for v in transformation[:4]
-            )
-            if transformation != (1, 0, 0, 1):
-                if (self.handleOverflowingTransforms and
-                        any(MAX_F2DOT14 < s <= 2 for s in transformation)):
-                    # clamp values ~= +2.0 so we can keep the component
-                    transformation = tuple(MAX_F2DOT14 if MAX_F2DOT14 < s <= 2
-                                           else s for s in transformation)
-                component.transform = (transformation[:2], transformation[2:])
-            component.flags = componentFlags
-            components.append(component)
-        return components
+    This pen can be used to construct or modify glyphs in a TrueType format
+    font. After using the pen to draw points, use the ``.glyph()`` method to
+    retrieve a :py:class:`~._g_l_y_f.Glyph` object representing the glyph.
+    """
 
-    def glyph(self, componentFlags=0x4):
-        """Returns a :py:class:`~._g_l_y_f.Glyph` object representing the glyph."""
-        assert self._isClosed(), "Didn't close last contour."
+    def _decompose(
+        self,
+        glyphName: str,
+        transformation: Tuple[float, float, float, float, float, float],
+    ):
+        tpen = TransformPointPen(self, transformation)
+        self.glyphSet[glyphName].drawPoints(tpen)
 
-        components = self._buildComponents(componentFlags)
+    def _isClosed(self) -> bool:
+        return self._currentContourStartIndex is None
 
-        glyph = Glyph()
-        glyph.coordinates = GlyphCoordinates(self.points)
-        glyph.coordinates.toInt()
-        glyph.endPtsOfContours = self.endPts
-        glyph.flags = array("B", self.types)
-        self.init()
+    def beginPath(
+        self, identifier: Optional[str] = None, **kwargs: Any
+    ) -> None:
+        """
+        Start a new sub path.
+        """
+        if not self._isClosed():
+            raise PenError("Didn't close previous contour.")
+        self._currentContourStartIndex = len(self.points)
 
-        if components:
-            glyph.components = components
-            glyph.numberOfContours = -1
+    def endPath(self) -> None:
+        """
+        End the current sub path.
+        """
+        # TrueType contours are always "closed"
+        if self._isClosed():
+            raise PenError("Contour is already closed.")
+        if self._currentContourStartIndex == len(self.points):
+            raise PenError("Tried to end an empty contour.")
```

Comment:
I wonder if it's too harsh to error out if the contour is empty. We can safely ignore it I guess, with no loss of information

Reply:
I agree beginning and immediately ending a path is "transparent", but then I wonder in which case a script may do it ... better to err on the safe side? Likewise, we could relax the checks about `endPath` being called again before a new `beginPath`, because `beginPath` may assume ending the previous path implicitly ... I think I'd leave it like it is.